rules/prefer-destructuring-in-parameters.js

```diff
@@ -0,0 +1,270 @@
+	};
+}
+
+function hasDirectiveInFunction(functionNode) {
+	const {body} = functionNode;
+	if (body.type !== 'BlockStatement') {
+		return false;
+	}
+
+	return body.body.some(({directive}) => directive === 'use strict');
+}
+
+const create = context => {
+	const {ecmaVersion} = context.parserOptions;
+	const sourceCode = context.getSourceCode();
+	return {
+		':function > Identifier.params'(parameter) {
+			const {name, parent: functionNode} = parameter;
+
+			// If "use strict" directive used, it should not reported
+			// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Strict_Non_Simple_Params
+			if (hasDirectiveInFunction(functionNode)) {
+				return;
+			}
+
+			const scope = context.getScope();
+			const variable = findVariable(scope, parameter);
+
+			/* istanbul ignore next: This should not happen, but integration test fails on some typescript project, and can't reproduce in tests */
+			if (!variable) {
+				return;
+			}
+
+			const identifiers = variable.references.map(({identifier}) => identifier);
+
+			const properties = new Map();
+			let propertyType;
+			let firstExpression;
+			for (const identifier of identifiers) {
+				const memberExpression = getMemberExpression(identifier);
+				if (!memberExpression) {
+					return;
+				}
+
+				const {node, type} = memberExpression;
+				if (propertyType) {
+					// Avoid case like `foo[0] === foo.length`
+					if (type !== propertyType) {
+						return;
+					}
+				} else {
+					propertyType = type;
+				}
+
+				if (
+					!firstExpression ||
+					node.range[0] < firstExpression.node.range[0]
+				) {
+					firstExpression = memberExpression;
+				}
+
+				const indexOrProperty = memberExpression[type];
+				if (properties.has(indexOrProperty)) {
+					properties.get(indexOrProperty).memberExpressions.push(node);
+				} else {
+					properties.set(indexOrProperty, {memberExpressions: [node]});
+				}
+			}
+
+			if (properties.size === 0) {
+				return;
+			}
+
+			const scopes = [
+				variable.scope,
+				...variable.references.map(({from}) => from)
+			];
+			for (const [indexOrProperty, data] of properties.entries()) {
+				let variableName;
+				if (propertyType === 'index') {
+					variableName = avoidCapture(
+						`${indexVariableNamePrefixes[indexOrProperty]}ElementOf${upperFirst(name)}`,
+						scopes,
+						ecmaVersion
+					);
+				} else {
+					variableName = avoidCapture(indexOrProperty, scopes, ecmaVersion);
+					if (variableName !== indexOrProperty) {
+						return;
+					}
+				}
+
+				data.variable = variableName;
+			}
+
+			context.report({
+				node: firstExpression.node,
+				messageId: MESSAGE_ID,
+				data: {
+					member: `${name}${propertyType === 'index' ? `[${firstExpression.index}]` : `.${firstExpression.property}`}`,
+					parameter: name
+				},
+				fix: fix({
+					sourceCode,
+					functionNode,
+					parameter,
+					properties,
+					type: propertyType
+				})
+			});
+		}
+	};
+};
+
+module.exports = {
```

Comment:
We'll want to add a `type` annotation here (more info in https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1606):

```suggestion
/** @type {import('eslint').Rule.RuleModule} */
module.exports = {
```